### PR TITLE
Sprint_1 first PR

### PR DESCRIPTION
**Pull request by PiePawPPI3**
Sprint_1

### CODE REVIEW

**Comment by xebia-pdalka** [comment]
`.idea` folder is never pushed to git

File: functions/__pycache__/__init__.cpython-311.pyc
Comment (xebia-pdalka) [inline functions/__pycache__/__init__.cpython-311.pyc:1]: `__pycache__` folders are never stored in git

File: functions/methods.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
+from typing import List
+
+
+def read_the_file(file_path: str):
```
Comment (xebia-pdalka) [inline functions/methods.py]: Please add returned type for all functions

File: functions/methods.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+        data = file.readlines()
+        if data:
+            print('file_content:')
+            for line in data:
+                process_the_content(line.strip())
+        else:
+            print('file_is_empty.')
+
+
+def calculate_average(scores: List[float]):
```
Comment (xebia-pdalka) [inline functions/methods.py]: There is no need to import common collections from `typing` - you can use directly `list[float]` here

File: main.py [hunk 3]
```diff
@@ -1,16 +1,11 @@
-# This is a sample Python script.
+# main.py
+from functions.methods import read_the_file
 
-# Press Shift+F10 to execute it or replace it with your code.
-# Press Double Shift to search everywhere for classes, files, tool windows, actions, and settings.
 
+def main():
+    file_path = "resources/grades1.txt"
+    read_the_file(file_path)
```
Comment (xebia-pdalka) [inline main.py]: Let's work further on single responsibility of functions.

- `read_the_file` should only read the file are return its content in the collection chosen by you
- then, there should be a function that parses data read, calculates averages and return them in one structure
- and finally there should be a function to print data (scores of all users and the average) taking as an input the results of the previous two functions

File: functions/methods.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+def print_user_average_score(username: str, average: float) -> None:
+    print(f'User {username} has average: {average:.2f}')
+
+
+def process_the_content(line: str) -> None:
+    fields = line.strip().split(',')
+    username = fields[0]
+    scores = [float(score) for score in fields[1:]
+              if score.strip().replace(',', '', 1).isdigit()]
```
Comment (xebia-pdalka) [inline functions/methods.py]: Let's try to use "duck typing" approach instead (`try: ... except` instead of custom conditions)

**Review by xebia-pdalka — CHANGES_REQUESTED** [review]

File: functions/methods.py (hunk 1 above)
Comment (PiePawPPI3) [reply functions/methods.py]: solved

File: functions/methods.py (hunk 2 above)
Comment (PiePawPPI3) [reply functions/methods.py]: done

File: functions/methods.py [hunk 5]
```diff
@@ -1,33 +1,50 @@
...
-def read_the_file(file_path: str):
-    with open(file_path, 'r') as file:
-        data = file.readlines()
-        if data:
-            print('file_content:')
-            for line in data:
-                process_the_content(line.strip())
-        else:
-            print('file_is_empty.')
+def read_the_file(file_path: str) -> list:
```
Comment (xebia-pdalka) [inline functions/methods.py]: `-> list` of what?

File: functions/methods.py [hunk 6]
```diff
@@ -1,33 +1,50 @@
...
+    for line in lines:
+        fields = line.strip().split(',')
+        username = fields[0]
+        scores = []
+        for score in fields:
+            try:
+                score_value = float(score.strip().replace(',', '', 1))
+                scores.append(score_value)
+            except ValueError:
+                pass
```
Comment (xebia-pdalka) [inline functions/methods.py]: This `for` loop for parsing input data should be moved to `read_the_file` as it covers file content validation and parsing into Python structures

File: functions/methods.py [hunk 7]
```diff
@@ -1,33 +1,50 @@
...
 
-def process_the_content(line: str) -> None:
-    fields = line.strip().split(',')
-    username = fields[0]
-    scores = [float(score) for score in fields[1:]
-              if score.strip().replace(',', '', 1).isdigit()]
-    average = calculate_average(scores)
 
-    print_user_average_score(username, average)
+print("*** Created at :", datetime.datetime.now().strftime('%Y-%m-%d %H:%M'))
```
Comment (xebia-pdalka) [inline functions/methods.py]: This should not be on module level

**Review by xebia-pdalka — CHANGES_REQUESTED** [review]
Let's refactor further:
- instead of one package `functions` create two packages `adapters` for input/output functionalities and `business_logic` for implementation of core behaviour.
- move functions from `functions` to an appropriate modules

File: functions/methods.py [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+        if not data:
+            print('File is empty.')
+        return data
+
+    except FileNotFoundError:
+        print('File not found.')
+
+
+def process_the_content(lines: list) -> list[dict]:
```
Comment (xebia-pdalka) [inline functions/methods.py]: Let's work on function names

File: adapters/data_parser.py [hunk 9]
```diff
@@ -0,0 +1,41 @@
+import datetime
+
+
+def read_scores(file_path: str) -> list[dict]:
+    try:
+        with open(file_path, 'r') as file:
+            data = file.readlines()
```
Comment (xebia-pdalka) [inline adapters/data_parser.py]: Here `try` block should end

File: adapters/data_parser.py [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+
+        if not data:
+            print('File is empty.')
```
Comment (xebia-pdalka) [inline adapters/data_parser.py]: printing should be moved to other place

File: adapters/data_parser.py [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+                    scores.append(score_value)
+                except ValueError:
+                    pass
+
+            results.append({'username': username, 'scores': scores})
+
+        return results
+
+    except FileNotFoundError:
+        print('File not found.')
```
Comment (xebia-pdalka) [inline adapters/data_parser.py]: As above

File: busines_logic/calculate_averages.py [hunk 12]
```diff
@@ -0,0 +1,15 @@
+def preparing_statistics(lines: list[dict]) -> list[dict]:
```
Comment (xebia-pdalka) [inline busines_logic/calculate_averages.py]: `prepare_statistics`

and type annotations: list of dict of what?

**Review by xebia-pdalka — APPROVED** [review]

**Pull request merged**